File: easy/validParens.py
class Solution(object):
    def isValid(self, s):
        """
        https://leetcode.com/problems/valid-parentheses/
        """
        stack = []
        for char in s:
            if char in '({[':
                stack.append(char)
            else:
                if len(stack) == 0:
                    print('invalid list, encountered closer with no opener', char)
                    return False
                top = stack[-1]
                if top == "(" and char == ")" \
                or top == "{" and char == "}" \
                or top == "[" and char == "]":
                    stack.pop()
                else:
                    print('invalid list, encountered closer with no opener', char)
                    return False
        if stack: 
            return False # if anything's left after looping through return false
        return True
        # O(n) time
        # O(n) space

File: easy/valid_parens.js
/**
 * @param {string} s
 * @return {boolean}
 */
const isValid = function(s) {
  // empty is valid? yes
  // will it have any other kind of characters?
  // how big???
  // loop through, adding opens to an array, when encountering a close,
  // it should have a corresponding open at the end of the array
  // don't need the index
  if (s.length % 2 !== 0) return false // if odd, incorrect
  const stack = []
  const openersMap = { "(":")", "{":"}", "[":"]"}
  for(let elem of s) {
      if (elem in openersMap) {
          stack.push(elem)
      } else {
          if (stack.length === 0) return false
          let stackTop = stack.pop()
          if (openersMap[stackTop] !== elem) {
              return false
          }
      }
  }
  return stack.length === 0
};
// O(n) time
// O(n) space - creating a stack with it will be half as long